docs/_posts/2022-10-31-luau-semantic-subtyping.md

```diff
@@ -0,0 +1,239 @@
+*  `V` = intersection of `{ p : c, q : c } -> never` for each vertex `(p, q)` and color `c` 
+*  `C` = intersection of `{ p : c }` for each node node `p` and color `c`
+
+Now `N & V` is a subtype of `C -> never` when it is impossible to color the graph, and so subtype checking is NP-hard. We deal with this in two ways:
+
+* we cache types to reduce memory footprint, and
+* give up with a “Code Too Complex” error if the cache of types gets too large.
+
+Hopefully this doesn’t come up in practice much. There is good evidence that issues like this don’t arise in practice from experience with type systems like that of Standard ML, which is EXPTIME-complete, but in practice you have to go out of your way to code up Turing Machine tapes as types.
+
+## Type normalization
+
+The algorithm used to decide semantic subtyping is *type normalization*.
+Rather than being directed by syntax, we first rewrite types to be normalized, then check subtyping on normalized types.
+
+A normalized type is a union of:
+
+* a normalized nil type (either `never` or `nil`)
+* a normalized number type (either `never` or `number`)
+* a normalized boolean type (either `never` or `true` or `false` or `boolean`)
+* a normalized function type (either `never` or an intersection of function types)
+etc
+
+Once types are normalized, it is straightforward to check semantic subtyping.
+
+Every type can be normalized (sigh, with some technical restrictions around generic type packs). The important steps are:
+
+* removing intersections of mismatched primitives, e.g. `number & bool` is replaced by `never`, and
+* removing unions of functions, e.g. `((number?) -> number) | ((string?) -> string)` is replaced by `(nil) -> (number | string)`.
+
+For example, normalizing `(number?) & (string?)` removes `number & string`, so all that is left is `nil`.
+
+Our first attempt at implementing type normalization applied it liberally, but this resulted in dreadful performance (complex code went from typechecking in less than a minute to running overnight). The reason for this is annoyingly simple: there is an optimization in Luau’s subtyping algorithm to handle reflexivity (`T` is a subtype of `T`) that performs a cheap pointer equality check. Type normalization can convert pointer-identical types into semantically-equivalent (but not pointer-identical) types, which significantly degrades performance.
+
+Because of these performance issues, we still use syntactic subtyping as our first check for subtyping, and only perform type normalization if the syntactic algorithm fails. This is sound, because syntactic subtyping is a conservative approximation to semantic subtyping.
+
+## Pragmatic semantic subtyping
+
+Off-the-shelf semantic subtyping is slightly different from what is implemented in Luau, because it requires models to be *set-theoretic*, which requires that inhabitants of function types “act like functions”. There are two reasons why we drop this requirement.
```

Comment:
```suggestion
Off-the-shelf semantic subtyping is slightly different from what is implemented in Luau, because it requires models to be *set-theoretic*, which requires that inhabitants of function types “act like functions.” There are two reasons why we drop this requirement.
```